### PR TITLE
Create the OpLike protocol and refactor Op | feat(values)

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #692
* #674
* #698

## Changes

- Creates the `OpLike` protocol that defines common attributes and methods for `Op`, `OnnxFunction` and `TraceOnlyFunction` so we can assume a common interface.
- Implement `param_schemas` for `TraceOnlyFunction`
- Refactor `param_schemas` to extract common logic.
- Removes `is_single_op` from `Op` because it is unused.

Signed-off-by: Justin Chu <justinchu@microsoft.com>

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/tracing.py [hunk 1]
```diff
@@ -33,7 +37,7 @@ def _get_src_and_ast(f: types.FunctionType) -> tuple[str, ast.FunctionDef]:
     return src, f_ast
 
 
-class TraceOnlyFunction:
+class TraceOnlyFunction(onnxscript.values.OpLike):
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/tracing.py:40]: What are the differences between a TraceOnlyFunction and an OnnxFunction? Eg., would it make sense for an OnnxFunction to be an extension (or derived class) of a TraceOnlyFunction?

**Review by titaiwangms — COMMENTED** [review]
I suggest we move `TraceOnlyFunction` to values.py. It would make more sense to the existence of OpLike. And if we decide to go with the current setting, we should change the file name tracing to something more specific. Tracing seems general... Codes are all good!

File: onnxscript/function_libs/torch_lib/tracing.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:40]: A TraceOnlyFunction is simply a python function with added attributes to make it more like an onnx function. To me it is a semantically different thing that happens to implement the same protocol for convenience. But I could be wroing

**Review by titaiwangms — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/tracing.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:40]: Let me see if a TraceOnlyFunction can be derived from Op

**Pull request closed without merging**